src/python_testing/TC_BRBINFO_4_1.py

```diff
@@ -0,0 +1,286 @@
+
+    #
+    # Test Harness Helpers
+    #
+
+    # Override default timeout to support a 60 min wait
+    @property
+    def default_timeout(self) -> int:
+        return 63*60
+
+    def desc_TC_BRBINFO_4_1(self) -> str:
+        """Returns a description of this test"""
+        return "[TC_BRBINFO_4_1] Verification of KeepActive Command [DUT-Server]"
+
+    def steps_TC_BRBINFO_4_1(self) -> list[TestStep]:
+        steps = [
+            TestStep("0", "Preconditions"),
+            TestStep("1a", "TH reads from the ICD the A_IDLE_MODE_DURATION, A_ACTIVE_MODE_DURATION, and ACTIVE_MODE_THRESHOLD attributes"),
+            TestStep("1b", "Simple KeepActive command w/ subscription. ActiveChanged event received by TH contains PromisedActiveDuration"),
+            TestStep("2", "Sends 3x KeepActive commands w/ subscription. ActiveChanged event received ONCE and contains PromisedActiveDuration"),
+            TestStep("3", "KeepActive not returned after 60 minutes of offline ICD"),
+        ]
+        return steps
+
+    def _ask_for_vendor_commissioniong_ux_operation(self, discriminator, setupPinCode, setupManualCode, setupQRCode):
+        self.wait_for_user_input(
+            prompt_msg=f"Using the DUT vendor's provided interface, commission the ICD device using the following parameters:\n"
+            f"- discriminator: {discriminator}\n"
+            f"- setupPinCode: {setupPinCode}\n"
+            f"- setupQRCode: {setupQRCode}\n"
+            f"- setupManualcode: {setupManualCode}\n"
+            f"If using FabricSync Admin test app, you may type:\n"
+            f">>> pairing onnetwork 111 {setupPinCode}")
+
+    async def _send_keep_active_command(self, duration, endpoint_id) -> int:
+        logging.info("Sending keep active command")
+        keep_active = await self.default_controller.SendCommand(nodeid=self.dut_node_id, endpoint=endpoint_id, payload=Clusters.Objects.BridgedDeviceBasicInformation.Commands.KeepActive(stayActiveDuration=duration))
+        return keep_active
+
+    async def _wait_for_active_changed_event(self, timeout) -> int:
+        try:
+            promised_active_duration = self.q.get(block=True, timeout=timeout)
+            logging.info(f"PromisedActiveDuration: {promised_active_duration}")
+            return promised_active_duration
+        except queue.Empty:
+            asserts.fail("Timeout on event ActiveChanged")
+
+    async def _get_dynamic_endpoint(self) -> int:
+        root_part_list = await self.read_single_attribute_check_success(cluster=Clusters.Descriptor, attribute=Clusters.Descriptor.Attributes.PartsList, endpoint=kRootEndpointId)
+        set_of_endpoints_after_adding_device = set(root_part_list)
+
+        asserts.assert_true(set_of_endpoints_after_adding_device.issuperset(
+            self.set_of_dut_endpoints_before_adding_device), "Expected only new endpoints to be added")
```

Comment:
nit,

recommend passing `self.set_of_dut_endpoints_before_adding_device` as a parameter instead and not adding it to self ever just having it be a local variable

Reply:
nvm I see what you did